changes/2020-06-04_how-to-fail-with-keyrings/change.md

```diff
@@ -0,0 +1,223 @@
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# How to fail with Keyrings in the AWS Encryption SDK
+
+## Affected Features
+
+This serves as a reference of all features that this change affects.
+
+| Feature                                                                                                |
+| ------------------------------------------------------------------------------------------------------ |
+| [Keyring Failure Communication](https://github.com/awslabs/aws-encryption-sdk-specification/issues/40) |
+
+## Affected Specifications
+
+This serves as a reference of all specification documents that this change affects.
+
+| Specification                                             |
+| --------------------------------------------------------- |
+| [Keyring Interface](../../framework/keyring-interface.md) |
+
+## Affected Implementations
+
+The scope of this change only affects the specification.
+Follow-up changes that define how this change
+affects specific keyring specifications
+will affect implementations.
+
+## Definitions
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+In order to define when keyrings MUST fail,
+we define a new concept of the "decryption contract",
+how it applies to keyrings generally,
+and how it is used to determine when keyrings MUST fail.
+
+## Out of Scope
+
+- The specifics of how ESDK components
+  communicate failure information between themselves
+  is out of scope.
+  We will address this in future specification changes.
+
+- The specific nature of _how_ ESDK framework components indicate failure
+  is out of scope.
+  That will depend on the best practices for each language.
+
+- Defining the decryption contract for each keyring
+  is out of scope.
+  We will address this in future specification changes.
+
+- A mechanism to interrogate a keyring
+  to determine the properties of its decryption contract
+  is out of scope.
+  We SHOULD address this in future changes.
+
+- This change only applies to keyrings.
+  We SHOULD define analogous concepts for
+  cryptographic materials managers
+  and the AWS Encryption SDK client
+  in future changes.
+
+## Motivation
+
+We lack a consistent definition both of
+what it means for a keyring to fail
+and what causes a keyring to fail.
+Without this consistent definition,
+different keyrings and implementations
+do slightly different things,
+leading to implementation drift
+and confusion for users and implementors.
+
+In order to clarify this situation,
+we MUST first define what it means for a keyring to fail,
+and that means defining what it means for a keyring to succeed.
+To do this,
+we define a new concept of the
+"keyring decryption contract".
```

Comment:
The "keyring" part here is still the main sticking point for me: is the decryption contract ultimately owned by the artifact, or is it attached to the keyring that produced it and the artifact inherits it?

I think this is the key question: if it is possible to encrypt using a Master Key and decrypt using a Keyring, does that mean the contract is ultimately a property of the artifact(s), and the "keyring decryption contract" should be "the decryption contract of the artifacts this keyring produces"? Otherwise we are stating that there is no contract involved in the migration path.

Reply:
The line I'm trying to thread here is that:

1. These artifacts are created and consumed by keyrings.
1. The keyring's decryption contract MUST be fulfilled through the use of that artifact.
1. In future changes we SHOULD introduce a concept of a CMM decryption contract, which is different than the decryption contracts that are written and fulfilled by keyrings.

The keyring is the actor, so "keyring decryption contract". Unless we can come up with a better name, I would be inclined to say that that CMM analogue should be a "CMM decryption contract."

Reply:
I see your point about disambiguating the scope of the contract. Would it be better to use "key decryption contract" instead?

Reply:
Why not "decryption contract" for the artifact and then it can be qualified by who it handling or verifying the contract?

Reply:
As in the decryption contract for the artifact is implicitly a "key decryption contract" because of the kind of artifact it is, and so the decryption contract for the whole message is implicitly a "CMM decryption contract" or similar?

Reply:
I personally prefer the `keyring decryption contract`. This is less about how to decrypt the key and more about what the keyring needs to be able to decrypt the key. For example, for the AWS KMS Keyring, you need AWS KMS Decrypt API permissions. The key itself doesn't care about network calls/ AWS permissions, but the keyring does care.